lib/rust/ensogl/core/src/display/render/passes/cache_shapes.rs

```diff
@@ -0,0 +1,373 @@
+    #[derivative(Debug = "ignore")]
+    shapes_to_render: Vec<Rc<dyn AnyShapeView>>,
+    texture_width:    i32,
+    texture_height:   i32,
+    layer:            Layer,
+}
+
+impl CacheShapesPass {
+    /// Constructor.
+    pub fn new(scene: &Scene) -> Self {
+        Self {
+            framebuffer:      default(),
+            shapes_to_render: default(),
+            layer:            Layer::new("Cached Shapes"),
+            scene:            scene.clone_ref(),
+            texture_width:    default(),
+            texture_height:   default(),
+        }
+    }
+}
+
+
+// === [`pass::Definition`] Implementation ===
+
+impl pass::Definition for CacheShapesPass {
+    fn initialize(&mut self, instance: &Instance) {
+        let ArrangedShapes { texture_width, texture_height, shapes } =
+            display::world::CACHED_SHAPES_DEFINITIONS
+                .with_borrow(|shapes| arrange_shapes_on_texture(shapes));
+        self.shapes_to_render = shapes.into_iter().map(Box::into).collect();
+        self.texture_width = texture_width;
+        self.texture_height = texture_height;
+
+        for shape in &self.shapes_to_render {
+            self.scene.add_child(&**shape);
+            self.layer.add(&**shape);
+        }
+        self.layer.camera().set_screen(texture_width as f32, texture_height as f32);
+        // We must call update of layer and display object hierarchy at this point, because:
+        // 1. the [`self.layer`] is not in the Layer hierarchy, so it's not updated during routine
+        //    layers update.
+        // 2. The pass can be re-initialized after the display object hierarchy update, but before
+        //    rendering, so the herarchy could be outdated during rendering.
+        self.layer.camera().update(&self.scene);
+        self.scene.display_object.update(&self.scene);
+        self.layer.update();
+
+        let output = pass::OutputDefinition::new_rgba("cached_shapes");
+        let texture = instance.new_texture(&output, texture_width, texture_height);
+        self.framebuffer = Some(instance.new_framebuffer(&[&texture]));
+    }
+
+    fn run(&mut self, instance: &Instance, _update_status: UpdateStatus) {
+        let is_shader_compiled =
+            |shape: &mut Rc<dyn AnyShapeView>| shape.sprite().symbol.shader().program().is_some();
+        let mut ready_to_render = self.shapes_to_render.drain_filter(is_shader_compiled).peekable();
+        if ready_to_render.peek().is_some() {
+            if let Some(framebuffer) = self.framebuffer.as_ref() {
+                framebuffer.bind();
+                instance.with_viewport(self.texture_width, self.texture_height, || {
+                    with_context(|ctx| ctx.set_camera(&self.layer.camera()));
+                    for shape in ready_to_render {
+                        shape.sprite().symbol.render();
+                    }
+                });
+                instance.context.bind_framebuffer(*Context::FRAMEBUFFER, None);
```

Comment:
can we replace it with something like `framebuffer.with(|| {...})` which will automatically set it to `None` at the end?